### PR TITLE
SONAR-8964 Drop overall notifications not related to the current user on SonarCloud

### PR DESCRIPTION
**Pull request by julienlancelot**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/notification/ws/AddAction.java [hunk 1]
```diff
@@ -45,26 +45,29 @@
...
 
 public class AddAction implements NotificationsWsAction {
   private final NotificationCenter notificationCenter;
   private final NotificationUpdater notificationUpdater;
+  private final Dispatchers dispatchers;
   private final DbClient dbClient;
   private final ComponentFinder componentFinder;
   private final UserSession userSession;
   private final List<String> globalDispatchers;
   private final List<String> projectDispatchers;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/notification/ws/AddAction.java]: Code Smell Code Smell: Remove this unused "projectDispatchers" private field. ([squid:S1068](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1068))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fjl%2FSONAR-8964_global_notif_on_sonarcloud&issues=AWFxXZ532VQm03fG7H_k&open=AWFxXZ532VQm03fG7H_k)

[ISSUE_KEY]: AWFxXZ532VQm03fG7H_k

File: server/sonar-server/src/main/java/org/sonar/server/notification/ws/AddAction.java [hunk 2]
```diff
@@ -45,26 +45,29 @@
...
+import static org.sonar.server.ws.WsUtils.checkRequest;
 
 public class AddAction implements NotificationsWsAction {
   private final NotificationCenter notificationCenter;
   private final NotificationUpdater notificationUpdater;
+  private final Dispatchers dispatchers;
   private final DbClient dbClient;
   private final ComponentFinder componentFinder;
   private final UserSession userSession;
   private final List<String> globalDispatchers;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/notification/ws/AddAction.java]: Code Smell Code Smell: Remove this unused "globalDispatchers" private field. ([squid:S1068](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1068))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fjl%2FSONAR-8964_global_notif_on_sonarcloud&issues=AWFxXZ532VQm03fG7H_j&open=AWFxXZ532VQm03fG7H_j)

[ISSUE_KEY]: AWFxXZ532VQm03fG7H_j

File: server/sonar-server/src/main/java/org/sonar/server/notification/ws/DispatchersImpl.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+import org.sonar.server.issue.notification.NewIssuesNotificationDispatcher;
+import org.sonar.server.notification.NotificationCenter;
+
+import static org.sonar.core.util.stream.MoreCollectors.toList;
+import static org.sonar.server.notification.NotificationDispatcherMetadata.GLOBAL_NOTIFICATION;
+import static org.sonar.server.notification.NotificationDispatcherMetadata.PER_PROJECT_NOTIFICATION;
+
+public class DispatchersImpl implements Dispatchers, Startable {
+
+  private static final Set<String> GLOBAL_DISPATCHERS_NOT_FOR_SONAR_CLOUD = ImmutableSet.of(
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/notification/ws/DispatchersImpl.java]: suggestion: rename to `GLOBAL_DISPATCHERS_TO_IGNORE_ON_SONAR_CLOUD`

File: server/sonar-server/src/main/java/org/sonar/server/notification/ws/DispatchersImpl.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  public List<String> getSortedProjectDispatchers() {
+    return projectDispatchers;
+  }
+
+  @Override
+  public void start() {
+    boolean isOnSonarCloud = configuration.getBoolean(ProcessProperties.Property.SONARCLOUD_ENABLED.getKey()).orElse(false);
+    this.globalDispatchers = notificationCenter.getDispatcherKeysForProperty(GLOBAL_NOTIFICATION, "true")
+      .stream()
+      .filter(dispatcher -> !(isOnSonarCloud && GLOBAL_DISPATCHERS_NOT_FOR_SONAR_CLOUD.contains(dispatcher)))
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/notification/ws/DispatchersImpl.java]: I'd create a Predicate for that

File: server/sonar-server/src/main/java/org/sonar/server/notification/ws/RemoveAction.java [hunk 5]
```diff
@@ -100,8 +97,8 @@ public void define(WebService.NewController context) {
         "  <li>Global notifications: %s</li>" +
         "  <li>Per project notifications: %s</li>" +
         "</ul>",
-        globalDispatchers.stream().sorted().collect(Collectors.joining(", ")),
-        projectDispatchers.stream().sorted().collect(Collectors.joining(", ")))
+        dispatchers.getSortedGlobalDispatchers().stream().sorted().collect(Collectors.joining(", ")),
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/notification/ws/RemoveAction.java]: minor: I'd rename to `getGlobalDispatchers()`

File: cix.sh [hunk 6]
```diff
@@ -43,7 +43,7 @@ case "$RUN_ACTIVITY" in
           ;;
 
         Category4)
-          CATEGORY="Category4|duplication"
```
Comment (teryk) [inline cix.sh]: what is this?

File: tests/src/test/java/org/sonarqube/tests/user/UserSuite.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+  BaseIdentityProviderTest.class,
+  FavoritesWsTest.class,
+  ForceAuthenticationTest.class,
+  LocalAuthenticationTest.class,
+  MyAccountPageTest.class,
+  OAuth2IdentityProviderTest.class,
+  RootUserInStandaloneModeTest.class,
+  UsersPageTest.class
+})
+public class UserSuite {
```
Comment (teryk) [inline tests/src/test/java/org/sonarqube/tests/user/UserSuite.java]: +💯

File: tests/src/test/java/org/sonarqube/tests/user/SonarCloudNotificationsWsTest.java [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    tester.as(user.getLogin()).wsClient().notifications().add(new AddRequest().setChannel("EmailNotificationChannel").setType("NewIssues").setProject(project.getKey()));
+
+    tester.as(user.getLogin()).wsClient().notifications().remove(new RemoveRequest().setChannel("EmailNotificationChannel").setType("ChangesOnMyIssue"));
+    tester.as(user.getLogin()).wsClient().notifications().remove(new RemoveRequest().setChannel("EmailNotificationChannel").setType("NewIssues").setProject(project.getKey()));
+
+    assertThat(tester.as(user.getLogin()).wsClient().notifications().list(new ListRequest()).getNotificationsList()).isEmpty();
+  }
+
+  @Test
+  public void fail_to_add_global_new_issues_notification() {
```
Comment (teryk) [inline tests/src/test/java/org/sonarqube/tests/user/SonarCloudNotificationsWsTest.java]: remark: I'd test it at unit level and not at IT level

File: server/sonar-server/src/main/java/org/sonar/server/notification/ws/DispatchersImpl.java (hunk 4 above)
Comment (julienlancelot) [reply server/sonar-server/src/main/java/org/sonar/server/notification/ws/DispatchersImpl.java]: ok

File: server/sonar-server/src/main/java/org/sonar/server/notification/ws/RemoveAction.java (hunk 5 above)
Comment (julienlancelot) [reply server/sonar-server/src/main/java/org/sonar/server/notification/ws/RemoveAction.java]: In fact it was my first intention but then I think that it's important to say that the list is sorted.

File: cix.sh (hunk 6 above)
Comment (julienlancelot) [reply cix.sh]: I can explain it to you

File: tests/src/test/java/org/sonarqube/tests/user/SonarCloudNotificationsWsTest.java (hunk 8 above)
Comment (julienlancelot) [reply tests/src/test/java/org/sonarqube/tests/user/SonarCloudNotificationsWsTest.java]: yes, and no... It's minor, but it's not costly so why not ?

**Pull request merged**